This notebook process staging dataset to targeted fact table.

In [148]:
import pandas as pd
import numpy as np
import yaml
from glob import glob
import os
from pathlib import Path
from datetime import date, timedelta, datetime
from pandas.tseries.holiday import USFederalHolidayCalendar
from pathlib import Path
import re
import hashlib
import calitp_data_analysis.magics
from google.cloud import bigquery

### Import staging dataset

In [149]:
df_dim = pd.read_csv("staging_ridership_output/dataset_metadata.csv")

In [150]:
df_dim.head()

,Unnamed: 0,dataset_id,dataset_name,organization_name,service_name,start_date_collected,end_date_collected,reporting_unit,ridership_measure,geographic_grain,...,stop_name_exists,stop_lat_exists,stop_lon_exists,boardings_exists,alightings_exists,ridership_exists,gtfs_url,gtfs_snapshot_date,gtfs_dataset_name,stop_coords_from_gtfs
0,0,189FC69D989010FD,golden_gate_park_shuttle_ridership,City and County of San Francisco,Golden Gate Park Shuttle,2024-07-01,2025-06-30,day,daily,stop,...,1,0,0,0,0,1,https://api.511.org/transit/datafeeds?operator...,2026-02-04,Bay Area 511 Golden Gate Park Shuttle Schedule,1.0
1,1,957BBF3AD8FC1B44,samtrans_ridership,San Mateo County Transit District,SamTrans,2025-08-01,2025-08-31,day,daily,trip_stop,...,1,1,1,1,1,0,https://api.511.org/transit/datafeeds?operator...,2026-02-04,Bay Area 511 SamTrans Schedule,1.0
2,2,1B1BADA1E00153AF,sunline_transit_ridership,SunLine Transit Agency,SunLine Transit,2022-07-01,2025-06-30,fiscal_year,avg_daily,stop,...,1,1,1,1,1,0,https://infopoint.sunline.org/InfoPoint/GTFS-Z...,2026-02-04,SunLine Avail Schedule,0.0
3,3,581003114DDAFDBE,santa_cruz_metro_ridership,Santa Cruz Metropolitan Transit District,Santa Cruz METRO,2024-07-01,2025-06-30,fiscal_year,total,stop,...,1,0,0,1,1,0,https://developer.scmetro.org/gtfs.zip,2026-02-04,Santa Cruz Schedule,1.0
4,4,1CCF9080DC015EB8,foothill_transit_ridership,Foothill Transit,Foothill Transit,2024-07-01,2025-06-30,day,daily,trip_stop,...,0,1,1,1,1,0,https://foothill3rdparty.rideralerts.com/mySto...,2026-02-04,Foothill Schedule,1.0


In [262]:
df_ridership = pd.read_csv("staging_ridership_output/staging_ridership_2026-03-11.csv", dtype={"stop_id": "string",
                                                                                                  "gtfs_stop_id": "string"})
df_ridership.head()

/tmp/ipykernel_288/2911636824.py:1: DtypeWarning: Columns (4,5,6,8) have mixed types. Specify dtype option on import or set low_memory=False.
  df_ridership = pd.read_csv("staging_ridership_output/staging_ridership_2026-03-11.csv", dtype={"stop_id": "string",


,record_id,dataset_id,organization_name,service_name,route_id,route_name,direction,stop_id,stop_name,stop_lat,stop_lon,daily_boardings,daily_alightings,daily_total_ridership,day_type,daily_ridership_basis,start_date,end_date,gtfs_stop_id
0,FF27A73278C43BCE,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,45.0,weekday,reported_daily,2024-07-01,2024-07-01,7602
1,31AC9C20F68713A9,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,59.0,weekday,reported_daily,2024-07-02,2024-07-02,7602
2,E021D40119E1054C,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,74.0,weekday,reported_daily,2024-07-03,2024-07-03,7602
3,B1F06206C22A5484,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,177.0,weekday,reported_daily,2024-07-04,2024-07-04,7602
4,5C7966570540D9F0,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,110.0,weekday,reported_daily,2024-07-05,2024-07-05,7602


In [263]:
len(df_ridership[(df_ridership["organization_name"] == "OmniTrans")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

1427

In [264]:
# take stop id provided by agency by default, if no stop id provided, use gtfs stop id
df_ridership["stop_id_comb"] = df_ridership["stop_id"].combine_first(df_ridership["gtfs_stop_id"])

In [265]:
# df_ridership[df_ridership["stop_id_comb"].isna()]["organization_name"].unique()

In [266]:
len(df_ridership[(df_ridership["organization_name"] == "OmniTrans")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

1427

**OmniTrans has stop names that are not matched in any stop name in GTFS.**

In [267]:
df_ridership[df_ridership["stop_id_comb"].isna()]

,record_id,dataset_id,organization_name,service_name,route_id,route_name,direction,stop_id,stop_name,stop_lat,stop_lon,daily_boardings,daily_alightings,daily_total_ridership,day_type,daily_ridership_basis,start_date,end_date,gtfs_stop_id,stop_id_comb
66381,84A0DE1B8E247D79,957BBF3AD8FC1B44,San Mateo County Transit District,SamTrans,NaN,NaN,NaN,<NA>,NaN,NaN,NaN,0.000000,0.000000,NaN,weekend,reported_daily,NaN,NaN,<NA>,<NA>
1387167,77E94F017028B7F4,40F5F343F0BD5850,OmniTrans,OmniTrans,1,NaN,NaN,<NA>,2ND @ F ST,NaN,NaN,2.565574,0.617486,NaN,all,calculated_avg_daily,2023-07-01,2024-06-30,<NA>,<NA>
1387168,DE8C7BC2A243DAF0,40F5F343F0BD5850,OmniTrans,OmniTrans,1,NaN,NaN,<NA>,2ND @ G ST,NaN,NaN,1.407104,7.680328,NaN,all,calculated_avg_daily,2023-07-01,2024-06-30,<NA>,<NA>
1387190,D6159BD22F54DE4A,40F5F343F0BD5850,OmniTrans,OmniTrans,1,NaN,NaN,<NA>,EAST VALLEY,NaN,NaN,4.169399,3.464481,NaN,all,calculated_avg_daily,2023-07-01,2024-06-30,<NA>,<NA>
1387191,DED32295F9E7560B,40F5F343F0BD5850,OmniTrans,OmniTrans,1,NaN,NaN,<NA>,GILBERT @ KENWOOD,NaN,NaN,0.707650,1.532787,NaN,all,calculated_avg_daily,2023-07-01,2024-06-30,<NA>,<NA>
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1391989,7229C9A19DCA034E,40F5F343F0BD5850,OmniTrans,OmniTrans,215,NaN,NaN,<NA>,EAST VALLEY,NaN,NaN,0.520548,0.416438,NaN,all,calculated_avg_daily,2025-07-01,2026-06-30,<NA>,<NA>
1391991,5D8E6C4CABCFEFE2,40F5F343F0BD5850,OmniTrans,OmniTrans,215,NaN,NaN,<NA>,MT VERNON @ CENTREPOINTE,NaN,NaN,16.873973,17.495890,NaN,all,calculated_avg_daily,2025-07-01,2026-06-30,<NA>,<NA>
1391994,430BAE15C3322FC7,40F5F343F0BD5850,OmniTrans,OmniTrans,290,NaN,NaN,<NA>,ARROWHEAD MEDICAL CENTER,NaN,NaN,0.958904,0.843836,NaN,all,calculated_avg_daily,2025-07-01,2026-06-30,<NA>,<NA>
1391995,2FE4D5BD7C13CAD8,40F5F343F0BD5850,OmniTrans,OmniTrans,290,NaN,NaN,<NA>,EAST VALLEY,NaN,NaN,0.063014,0.057534,NaN,all,calculated_avg_daily,2025-07-01,2026-06-30,<NA>,<NA>


### Handle OmniTrans Stop Name Format Inconsistency

In [268]:
# omni_mask = df_ridership["organization_name"] == "OmniTrans"
# t_df_omni = df_ridership.loc[omni_mask, ["stop_id_comb", "stop_name"]].copy()
# t_df_omni["_orig_index"] = t_df_omni.index

In [269]:
# street_types = ["street", "st", "avenue", "ave", "road", "rd", "boulevard", "blvd", "drive", "dr",
#                 "lane", "ln", "court", "ct", "place", "pl", "highway", "hwy", "way", "parkway", "pkwy"]

# street_type_regex = r"\b(?:" + "|".join(street_types) + r")\b"

In [270]:
# t_df_omni["stop_name_norm"] = t_df_omni["stop_name"].astype("string").str.lower().str.replace(r"\s+", " ", regex=True).str.strip()
# # detect presence of any street type
# t_df_omni["_has_street_type"] = t_df_omni["stop_name_norm"].str.contains(street_type_regex, regex=True, na=False)

# t_df_omni["_stop_name_length"] = t_df_omni["stop_name_norm"].str.len()
# # rank stop name variants within each stop id
# canonical = t_df_omni.sort_values(by=["_has_street_type", "_stop_name_length"],
#                                  ascending=[False, False]) \
#                     .groupby("stop_id_comb", as_index=False).first()[["stop_id_comb", "stop_name"]].rename(columns={"stop_name": "canonical_stop_name"})
# t_df_omni = t_df_omni.merge(canonical, on="stop_id_comb", how="left")
# canonical.head()

In [271]:
# df_ridership.loc[t_df_omni["_orig_index"], "stop_name"] = t_df_omni["canonical_stop_name"].values

In [272]:
# len(df_ridership[(df_ridership["organization_name"] == "OmniTrans")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

### Agg to get avg daily ridership and make sure there is only one single row for one stop: 

- filter data for only most recent year/period
- recalculate avg daily for each stop across the most recent year/period

use stop_id_comb or gtfs_stop_id

#### For the reported daily ridership (or transaction level)
- first agg to stop level (if raw is trip-stop, or transaction)
- then take avg across the time span to get avg daily ridership

In [273]:
df_ridership.head()

,record_id,dataset_id,organization_name,service_name,route_id,route_name,direction,stop_id,stop_name,stop_lat,stop_lon,daily_boardings,daily_alightings,daily_total_ridership,day_type,daily_ridership_basis,start_date,end_date,gtfs_stop_id,stop_id_comb
0,FF27A73278C43BCE,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,45.0,weekday,reported_daily,2024-07-01,2024-07-01,7602,7602
1,31AC9C20F68713A9,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,59.0,weekday,reported_daily,2024-07-02,2024-07-02,7602,7602
2,E021D40119E1054C,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,74.0,weekday,reported_daily,2024-07-03,2024-07-03,7602,7602
3,B1F06206C22A5484,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,177.0,weekday,reported_daily,2024-07-04,2024-07-04,7602,7602
4,5C7966570540D9F0,189FC69D989010FD,City and County of San Francisco,Golden Gate Park Shuttle,NaN,NaN,NaN,<NA>,Blue Heron Boathouse,37.77084,-122.476829,NaN,NaN,110.0,weekday,reported_daily,2024-07-05,2024-07-05,7602,7602


In [274]:
len(df_ridership["organization_name"].unique())

19

In [275]:
df_ridership["daily_ridership_basis"].unique()

array(['reported_daily', 'reported_avg_daily', 'calculated_avg_daily',
       'derived_from_transactions'], dtype=object)

In [276]:
df_ridership[["organization_name", "daily_ridership_basis"]].drop_duplicates().sort_values(by="daily_ridership_basis")

,organization_name,daily_ridership_basis
68672,Santa Cruz Metropolitan Transit District,calculated_avg_daily
1391999,Santa Barbara Metropolitan Transit District,calculated_avg_daily
1387167,OmniTrans,calculated_avg_daily
1405838,Riverside Transit Agency,derived_from_transactions
1366783,City of Santa Monica,reported_avg_daily
1404676,City of Culver City,reported_avg_daily
1400629,Sacramento Regional Transit District,reported_avg_daily
1399583,Gold Coast Transit District,reported_avg_daily
1377946,Long Beach Transit,reported_avg_daily
1883101,Peninsula Corridor Joint Powers Board,reported_avg_daily


In [277]:
df_ridership[df_ridership["daily_ridership_basis"] == "derived_from_transactions"]["organization_name"].unique()

array(['Riverside Transit Agency'], dtype=object)

In [278]:
df_ridership["organization_name"].unique()

array(['City and County of San Francisco',
       'San Mateo County Transit District', 'SunLine Transit Agency',
       'Santa Cruz Metropolitan Transit District', 'Foothill Transit',
       'City of Fresno', 'San Diego Metropolitan Transit System',
       'Golden Gate Bridge, Highway and Transportation District',
       'San Francisco Bay Area Rapid Transit District',
       'City of Santa Monica', 'Long Beach Transit', 'OmniTrans',
       'Santa Barbara Metropolitan Transit District',
       'Gold Coast Transit District',
       'Sacramento Regional Transit District', 'City of Culver City',
       'Riverside Transit Agency',
       'Orange County Transportation Authority',
       'Peninsula Corridor Joint Powers Board'], dtype=object)

In [279]:
df_reported_daily = df_ridership[(df_ridership["daily_ridership_basis"] == "reported_daily")].copy()
df_reported_daily["organization_name"].unique()

array(['City and County of San Francisco',
       'San Mateo County Transit District', 'Foothill Transit',
       'City of Fresno',
       'Golden Gate Bridge, Highway and Transportation District',
       'San Francisco Bay Area Rapid Transit District',
       'Orange County Transportation Authority'], dtype=object)

In [280]:
len(df_reported_daily["organization_name"].unique())

7

In [281]:
df_reported_daily = df_reported_daily.groupby(["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name",
                               "stop_lat", "stop_lon", "day_type", "daily_ridership_basis", "start_date", "end_date"], as_index=False, dropna=False) \
                    .agg(daily_total_ridership = ("daily_total_ridership", lambda x: x.sum(min_count=1)),
                         daily_alightings = ("daily_alightings", lambda x: x.sum(min_count=1)),
                         daily_boardings = ("daily_boardings", lambda x: x.sum(min_count=1))).reset_index()
df_reported_daily.head()

,index,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,day_type,daily_ridership_basis,start_date,end_date,daily_total_ridership,daily_alightings,daily_boardings
0,0,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,holiday,reported_daily,2024-10-14,2024-10-14,NaN,3666.0,3735.0
1,1,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,holiday,reported_daily,2024-11-11,2024-11-11,NaN,2301.0,2540.0
2,2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,holiday,reported_daily,2024-11-28,2024-11-28,NaN,976.0,1021.0
3,3,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,holiday,reported_daily,2024-11-29,2024-11-29,NaN,1883.0,1971.0
4,4,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,holiday,reported_daily,2024-12-24,2024-12-24,NaN,1931.0,1940.0


In [282]:
df_reported_daily["organization_name"].unique()

array(['San Francisco Bay Area Rapid Transit District',
       'City and County of San Francisco', 'Foothill Transit',
       'Orange County Transportation Authority',
       'San Mateo County Transit District', 'City of Fresno',
       'Golden Gate Bridge, Highway and Transportation District'],
      dtype=object)

In [283]:
len(df_reported_daily["organization_name"].unique())

7

In [284]:
# check if it is true that each stop has one record for each day (already checked all start date = end date)
df_reported_daily.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "start_date", "day_type"], as_index=False).size().sort_values(by="size", ascending=False).head(3)

,organization_name,stop_id,gtfs_stop_id,stop_name,start_date,day_type,size
0,City of Fresno,10,10,SHAW - CEDAR,2024-09-01,weekend,1
744556,Foothill Transit,2242,2242-a1,Reservoir St and Franklin Ave W,2024-07-07,weekend,1
744562,Foothill Transit,2242,2242-a1,Reservoir St and Franklin Ave W,2024-07-26,weekday,1


In [285]:
df_reported_daily_agg = df_reported_daily.groupby(by=["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name", 
                                                    "stop_lat", "stop_lon", "day_type"], as_index=False, dropna=False).agg(avg_daily_boardings = ("daily_boardings", "mean"),
                                                                                                            avg_daily_alightings = ("daily_alightings", "mean"),
                                                                                                            avg_daily_ridership = ("daily_total_ridership", "mean"),
                                                                                                            start_date = ("start_date", "min"),
                                                                                                            end_date = ("end_date", "max"))
df_reported_daily_agg.head()

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,day_type,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership,start_date,end_date
0,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,holiday,2120.214286,2108.857143,NaN,2024-10-14,2025-09-01
1,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,saturday,2681.788462,2603.000000,NaN,2024-10-05,2025-09-27
2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,sunday,1817.115385,1814.557692,NaN,2024-10-06,2025-09-28
3,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,weekday,5477.064777,5344.149798,NaN,2024-10-01,2025-09-30
4,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900209,19th Street / Oakland,37.808078,-122.268758,holiday,1979.714286,1905.000000,NaN,2024-10-14,2025-09-01


In [286]:
# df_reported_daily_agg = pd.merge(df_reported_daily_agg, df_dim[["dataset_id", "start_date_collected", "end_date_collected"]], on="dataset_id")
# df_reported_daily_agg.head()

In [287]:
df_reported_daily_agg["avg_daily_method"] = "avg_from_daily"

In [288]:
len(df_reported_daily_agg["organization_name"].unique())

7

In [289]:
# check duplicates
t_dup_check = df_reported_daily_agg.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False).size().sort_values(by="size", ascending=False)
t_dup_check[t_dup_check["size"] > 1].sort_values(by="stop_id")

,organization_name,stop_id,gtfs_stop_id,stop_name,day_type,size


In [290]:
# df_reported_daily_agg[(df_reported_daily_agg["organization_name"] == "Foothill Transit") &
#                       (df_reported_daily_agg["stop_id"] == "177") & 
#                       (df_reported_daily_agg["day_type"] == "weekday")]

In [291]:
# df_reported_daily_agg = df_reported_daily_agg.rename(columns={"start_date_collected": "start_date",
#                                          "end_date_collected": "end_date"})
# df_reported_daily_agg.head()

#### For Riverside (raw data is transaction data)

Raw data only provide stop id and doesn't look consistent with GTFS ids. And for each stop id, there can be multiple routes and multiple pairs of lat/lon. When doing the last step agg for each stop, we take the max lat and lon for each raw stop id for now. Can be improved in the future when we have better understanding of the data (especially what stop ID means).

In [292]:
df_riverside = df_ridership[df_ridership["organization_name"] == "Riverside Transit Agency"]
df_riverside.head(3)

,record_id,dataset_id,organization_name,service_name,route_id,route_name,direction,stop_id,stop_name,stop_lat,stop_lon,daily_boardings,daily_alightings,daily_total_ridership,day_type,daily_ridership_basis,start_date,end_date,gtfs_stop_id,stop_id_comb
1405838,65C5611E9B2393B4,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,1,NaN,Inbound,0,NaN,33.982152,-117.338144,NaN,NaN,69.0,weekday,derived_from_transactions,2025-05-01,2025-05-01,<NA>,0
1405839,C0608CC9B04359AA,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,3,NaN,Inbound,0,NaN,33.881796,-117.564320,NaN,NaN,9.0,weekday,derived_from_transactions,2025-05-01,2025-05-01,<NA>,0
1405840,7BCC12C0A0EE6511,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,3,NaN,Outbound,0,NaN,33.994648,-117.554640,NaN,NaN,2.0,weekday,derived_from_transactions,2025-05-01,2025-05-01,<NA>,0


In [293]:
df_riverside_agg = df_riverside.groupby(by=["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False).agg(avg_daily_boardings = ("daily_boardings", "mean"),
                                                                                                            avg_daily_alightings = ("daily_alightings", "mean"),
                                                                                                            avg_daily_ridership = ("daily_total_ridership", "mean"),
                                                                                                            stop_lat = ("stop_lat", "max"),
                                                                                                            stop_lon = ("stop_lon", "max"),
                                                                                                            start_date = ("start_date", "min"),
                                                                                                            end_date = ("end_date", "max"))
df_riverside_agg.head()

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,day_type,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership,stop_lat,stop_lon,start_date,end_date
0,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,0,<NA>,NaN,holiday,NaN,NaN,13.729730,33.994664,0.000000,2025-01-20,2025-10-13
1,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,0,<NA>,NaN,weekday,NaN,NaN,19.995297,34.226968,0.000000,2025-01-02,2025-10-31
2,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,0,<NA>,NaN,weekend,NaN,NaN,13.994141,34.484160,0.000000,2025-01-04,2025-10-26
3,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,1004,1004,NaN,holiday,NaN,NaN,34.666667,33.895190,-117.178521,2025-01-20,2025-10-13
4,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,1004,1004,NaN,weekday,NaN,NaN,46.886486,33.895190,-117.178521,2025-01-02,2025-10-31


In [294]:
df_riverside_agg["avg_daily_method"] = "avg_from_daily"

In [295]:
# check duplicates
t_dup_check = df_riverside_agg.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False).size().sort_values(by="size", ascending=False)
t_dup_check[t_dup_check["size"] > 1].sort_values(by="stop_id")

,organization_name,stop_id,gtfs_stop_id,stop_name,day_type,size


#### For the reported avg ridership

In [296]:
df_reported_avg = df_ridership[(df_ridership["daily_ridership_basis"] == "reported_avg_daily")].copy()
df_reported_avg["organization_name"].unique()

array(['SunLine Transit Agency', 'San Diego Metropolitan Transit System',
       'City of Santa Monica', 'Long Beach Transit',
       'Gold Coast Transit District',
       'Sacramento Regional Transit District', 'City of Culver City',
       'Peninsula Corridor Joint Powers Board'], dtype=object)

In [297]:
len(df_reported_avg["organization_name"].unique())

8

#### SDMTS, Long Beach Transit, provided avg daily for one year/fiscal year
#### Gold Coast Transit District, raw is avg for the given month (only one month worth of data)
#### SacRT raw is avg for given period (Sep-Dec 2023)

#### SunLine

raw data is avg for each fiscal year for 3 fiscal years, plus 2025/3 to 2025/5. We can keep most recent fiscal year.

In [298]:
df_reported_avg[(df_reported_avg["organization_name"] == "SunLine Transit Agency")][["start_date", "end_date"]].drop_duplicates()

,start_date,end_date
66382,2022-07-01,2023-06-30
66383,2023-07-01,2024-06-30
66384,2024-07-01,2025-06-30
66385,2025-03-01,2025-05-31


In [299]:
df_reported_avg = df_reported_avg[~((df_reported_avg["organization_name"] == "SunLine Transit Agency") & 
                                    (df_reported_avg["start_date"] != "2024-07-01"))]


In [300]:
df_reported_avg[(df_reported_avg["organization_name"] == "SunLine Transit Agency")]["start_date"].unique()

array(['2024-07-01'], dtype=object)

#### Santa Monica

raw data is avg for every 4 months from 2024-08-01 to 2025-11-30, 4 periods in total. Keep only most recent period.

In [301]:
df_reported_avg[(df_reported_avg["organization_name"] == "City of Santa Monica")]["start_date"].unique()

array(['2024-08-01', '2024-12-01', '2025-04-01', '2025-08-01'],
      dtype=object)

In [302]:
df_reported_avg[(df_reported_avg["organization_name"] == "City of Santa Monica")].head()

,record_id,dataset_id,organization_name,service_name,route_id,route_name,direction,stop_id,stop_name,stop_lat,stop_lon,daily_boardings,daily_alightings,daily_total_ridership,day_type,daily_ridership_basis,start_date,end_date,gtfs_stop_id,stop_id_comb
1366783,A804D30789A774D2,395857DB9F379E73,City of Santa Monica,Big Blue Bus,1,Main St & Santa Monica Blvd/UCLA,EASTBOUND,-4,MAIN NB/WESTMINSTER FS,33.990316,-118.472653,28.909881,0.111111,NaN,saturday,reported_avg_daily,2024-08-01,2024-11-30,346,-4
1366784,FDA60BBC003F98D9,395857DB9F379E73,City of Santa Monica,Big Blue Bus,1,Main St & Santa Monica Blvd/UCLA,EASTBOUND,1000,4TH NB/COLORADO FS (Downtown SM Station),34.014235,-118.492457,116.886196,187.964533,NaN,saturday,reported_avg_daily,2024-08-01,2024-11-30,21,1000
1366785,8A48393FDA926F9C,395857DB9F379E73,City of Santa Monica,Big Blue Bus,1,Main St & Santa Monica Blvd/UCLA,EASTBOUND,1054,4TH NB/PICO FS,34.010034,-118.487580,26.176476,41.298676,NaN,saturday,reported_avg_daily,2024-08-01,2024-11-30,284,1054
1366786,0BF5B3C802D0F7CB,395857DB9F379E73,City of Santa Monica,Big Blue Bus,1,Main St & Santa Monica Blvd/UCLA,EASTBOUND,1218,MAIN NB/HILL NS,34.000950,-118.482391,36.858219,4.749936,NaN,saturday,reported_avg_daily,2024-08-01,2024-11-30,340,1218
1366787,684FA140C497473F,395857DB9F379E73,City of Santa Monica,Big Blue Bus,1,Main St & Santa Monica Blvd/UCLA,EASTBOUND,1221,MAIN NB/MARINE NS,33.997991,-118.479444,73.455075,9.204640,NaN,saturday,reported_avg_daily,2024-08-01,2024-11-30,202,1221


In [303]:
df_reported_avg = df_reported_avg[~((df_reported_avg["organization_name"] == "City of Santa Monica") & 
                                    (df_reported_avg["start_date"] < df_reported_avg[df_reported_avg["organization_name"] == "City of Santa Monica"]["start_date"].max()))]


In [304]:
df_reported_avg[(df_reported_avg["organization_name"] == "City of Santa Monica")]["start_date"].unique()

array(['2025-08-01'], dtype=object)

#### Caltrain

raw data is avg for each month, from 2023-11-01 to 2025-07-01. Keep only most recent time period.

In [305]:
df_reported_avg[(df_reported_avg["organization_name"] == "Peninsula Corridor Joint Powers Board")]["start_date"].unique()

array(['2025-07-01', '2025-06-01', '2025-05-01', '2025-04-01',
       '2025-03-01', '2025-02-01', '2025-01-01', '2024-12-01',
       '2024-11-01', '2024-10-01', '2024-09-01', '2024-08-01',
       '2024-07-01', '2024-06-01', '2024-05-01', '2024-04-01',
       '2024-03-01', '2024-02-01', '2024-01-01', '2023-12-01',
       '2023-11-01'], dtype=object)

In [306]:
df_reported_avg[df_reported_avg["organization_name"] == "Peninsula Corridor Joint Powers Board"]["start_date"].max()

'2025-07-01'

In [307]:
df_reported_avg = df_reported_avg[~((df_reported_avg["organization_name"] == "Peninsula Corridor Joint Powers Board") & 
                                    (df_reported_avg["start_date"] < df_reported_avg[df_reported_avg["organization_name"] == "Peninsula Corridor Joint Powers Board"]["start_date"].max()))]


In [308]:
df_reported_avg[(df_reported_avg["organization_name"] == "Peninsula Corridor Joint Powers Board")]["start_date"].unique()

array(['2025-07-01'], dtype=object)

**For these agencies, just agg over routes and directions, no need to take avg again.**

In [309]:
df_reported_avg_agg = df_reported_avg.groupby(["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name",
                               "stop_lat", "stop_lon", "day_type", "start_date", "end_date"], as_index=False, dropna=False) \
                    .agg(avg_daily_boardings = ("daily_boardings", lambda x: x.sum(min_count=1)),
                         avg_daily_alightings = ("daily_alightings", lambda x: x.sum(min_count=1)),
                         avg_daily_ridership = ("daily_total_ridership", lambda x: x.sum(min_count=1))
                         ).reset_index()

In [310]:
df_reported_avg_agg.head()

,index,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,day_type,start_date,end_date,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership
0,0,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,1,1,Palm@2Bunch,33.947457,-116.501166,all,2024-07-01,2025-06-30,21.05,24.73,NaN
1,1,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,10,10,PalmCyn@Escuela,33.848167,-116.547967,all,2024-07-01,2025-06-30,1.50,0.85,NaN
2,2,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,100,100,Hwy111@Monroe,33.714485,-116.234272,all,2024-07-01,2025-06-30,52.39,28.68,NaN
3,3,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,101,101,Hwy111@LasPalms,33.711562,-116.238122,all,2024-07-01,2025-06-30,9.21,4.28,NaN
4,4,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,102,102,Hwy111@Clinton,33.707564,-116.245842,all,2024-07-01,2025-06-30,17.86,11.65,NaN


In [311]:
df_reported_avg_agg["avg_daily_method"] = "reported_avg_daily"

In [312]:
len(df_reported_avg_agg["organization_name"].unique())

8

In [313]:
# check duplicates
t_dup_check = df_reported_avg_agg.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False).size().sort_values(by="size", ascending=False)
t_dup_check[t_dup_check["size"] > 1]

,organization_name,stop_id,gtfs_stop_id,stop_name,day_type,size


#### For calculated_avg

- Santa Cruz Metro: raw is total for one fiscal year (2024-07-01 to 2025-06-30)
- OmniTrans: raw is total for each fiscal year from 2024-2026. Use fiscal year 2025 (2024-07-01 to 2025-06-30)
- SBMTD: raw is monthly total for each month from 2024-11-01 to 2025-10-31

Note: for now, just use # of days in the given year/month to calculate avg, and keep only the most recent time period.

TODO: in pre-process stage, bring in gtfs data to best count the number of service dates for ideally each stop (otherwise each agency)

In [314]:
df_calc_avg = df_ridership[(df_ridership["daily_ridership_basis"] == "calculated_avg_daily")].copy()
df_calc_avg["organization_name"].unique()

array(['Santa Cruz Metropolitan Transit District', 'OmniTrans',
       'Santa Barbara Metropolitan Transit District'], dtype=object)

In [315]:
df_calc_avg[df_calc_avg["organization_name"] == "OmniTrans"]["start_date"].unique()

array(['2023-07-01', '2024-07-01', '2025-07-01'], dtype=object)

In [316]:
df_calc_avg = df_calc_avg[~((df_calc_avg["organization_name"] == "OmniTrans") & 
                            (df_calc_avg["start_date"] != "2024-07-01"))]


In [317]:
df_calc_avg[df_calc_avg["organization_name"] == "OmniTrans"]["start_date"].unique()

array(['2024-07-01'], dtype=object)

In [318]:
df_calc_avg[df_calc_avg["organization_name"] == "Santa Barbara Metropolitan Transit District"]["start_date"].unique()

array(['2024-11-01', '2024-12-01', '2025-01-01', '2025-02-01',
       '2025-03-01', '2025-04-01', '2025-05-01', '2025-06-01',
       '2025-07-01', '2025-08-01', '2025-09-01', '2025-10-01'],
      dtype=object)

In [319]:
df_calc_avg = df_calc_avg[~((df_calc_avg["organization_name"] == "Santa Barbara Metropolitan Transit District") & 
                            (df_calc_avg["start_date"] < df_calc_avg[df_calc_avg["organization_name"] == "Santa Barbara Metropolitan Transit District"]["start_date"].max()))]


In [320]:
df_calc_avg[df_calc_avg["organization_name"] == "Santa Barbara Metropolitan Transit District"]["start_date"].unique()

array(['2025-10-01'], dtype=object)

**Then agg over routes and directions, no need to take avg again.**

In [321]:
df_calc_avg_agg = df_calc_avg.groupby(["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name",
                               "stop_lat", "stop_lon", "day_type", "start_date", "end_date"], as_index=False, dropna=False) \
                    .agg(avg_daily_boardings = ("daily_boardings", lambda x: x.sum(min_count=1)),
                         avg_daily_alightings = ("daily_alightings", lambda x: x.sum(min_count=1)),
                         avg_daily_ridership = ("daily_total_ridership", lambda x: x.sum(min_count=1))
                         ).reset_index()

In [322]:
df_calc_avg_agg["avg_daily_method"] = "avg_from_total"

In [323]:
# check duplicates
t_dup_check = df_calc_avg_agg.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False).size().sort_values(by="size", ascending=False)
t_dup_check[t_dup_check["size"] > 1]

,organization_name,stop_id,gtfs_stop_id,stop_name,day_type,size


In [324]:
# df_agg = df_ridership.groupby(["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name",
#                                "stop_lat", "stop_lon", "day_type", "daily_ridership_basis", "start_date", "end_date"], as_index=False, dropna=False) \
#                     .agg(daily_total_ridership = ("daily_total_ridership", lambda x: x.sum(min_count=1)),
#                          daily_alightings = ("daily_alightings", lambda x: x.sum(min_count=1)),
#                          daily_boardings = ("daily_boardings", lambda x: x.sum(min_count=1))).reset_index()

# # drop index col
# df_agg.drop(columns="index", inplace=True)

# df_agg.head()

**Note: The duplicates in SamTrans is the stop id that is not in GTFS - and in ridership data, a stop id can have more than one pair of lat/lon. Looks like it is vehicle position. So we use GTFS lat/lon mapped in using stop id for the aggregation.**

In [325]:
# df_org_res = df_org.groupby(["dataset_id", "organization_name", "service_name", "gtfs_stop_id", 
#                              "stop_id", "stop_name", "stop_lat", "stop_lon"], as_index=False).agg()

#### Combine the sub datasets (reported daily, reported avg, and calculated avg)

In [326]:
df_reported_daily_agg.head(3)

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,day_type,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership,start_date,end_date,avg_daily_method
0,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,holiday,2120.214286,2108.857143,NaN,2024-10-14,2025-09-01,avg_from_daily
1,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,saturday,2681.788462,2603.000000,NaN,2024-10-05,2025-09-27,avg_from_daily
2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.27163,sunday,1817.115385,1814.557692,NaN,2024-10-06,2025-09-28,avg_from_daily


In [327]:
df_riverside_agg.head(3)

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,day_type,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership,stop_lat,stop_lon,start_date,end_date,avg_daily_method
0,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,0,<NA>,NaN,holiday,NaN,NaN,13.729730,33.994664,0.0,2025-01-20,2025-10-13,avg_from_daily
1,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,0,<NA>,NaN,weekday,NaN,NaN,19.995297,34.226968,0.0,2025-01-02,2025-10-31,avg_from_daily
2,52A9DCDCDBA1CCD3,Riverside Transit Agency,Riverside Transit,0,<NA>,NaN,weekend,NaN,NaN,13.994141,34.484160,0.0,2025-01-04,2025-10-26,avg_from_daily


In [328]:
df_reported_avg_agg.head(3)

,index,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,day_type,start_date,end_date,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership,avg_daily_method
0,0,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,1,1,Palm@2Bunch,33.947457,-116.501166,all,2024-07-01,2025-06-30,21.05,24.73,NaN,reported_avg_daily
1,1,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,10,10,PalmCyn@Escuela,33.848167,-116.547967,all,2024-07-01,2025-06-30,1.50,0.85,NaN,reported_avg_daily
2,2,1B1BADA1E00153AF,SunLine Transit Agency,SunLine Transit,100,100,Hwy111@Monroe,33.714485,-116.234272,all,2024-07-01,2025-06-30,52.39,28.68,NaN,reported_avg_daily


In [329]:
df_calc_avg_agg.head(3)

,index,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,day_type,start_date,end_date,avg_daily_boardings,avg_daily_alightings,avg_daily_ridership,avg_daily_method
0,0,165BA8D319143274,Santa Barbara Metropolitan Transit District,Santa Barbara Metropolitan Transit District,1,1,Modoc & Portesuello,34.424858,-119.726070,all,2025-10-01,2025-10-31,48.000000,27.548387,75.516129,avg_from_total
1,1,165BA8D319143274,Santa Barbara Metropolitan Transit District,Santa Barbara Metropolitan Transit District,10,10,Anapamu & Santa Barbara,34.425301,-119.702300,all,2025-10-01,2025-10-31,7.000000,2.000000,9.000000,avg_from_total
2,2,165BA8D319143274,Santa Barbara Metropolitan Transit District,Santa Barbara Metropolitan Transit District,100,100,San Andres & Sola,34.417937,-119.714983,all,2025-10-01,2025-10-31,14.322581,32.096774,46.419355,avg_from_total


In [330]:
cols = ["dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name", "stop_lat", "stop_lon", 
        "avg_daily_alightings", "avg_daily_boardings", "avg_daily_ridership", "start_date", "end_date", "day_type", "avg_daily_method"]

In [331]:
df_agg_combined = pd.concat([df_reported_daily_agg[cols], df_riverside_agg[cols], df_reported_avg_agg[cols], df_calc_avg_agg[cols]], ignore_index=True)
df_agg_combined.head()

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,avg_daily_alightings,avg_daily_boardings,avg_daily_ridership,start_date,end_date,day_type,avg_daily_method
0,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2108.857143,2120.214286,NaN,2024-10-14,2025-09-01,holiday,avg_from_daily
1,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2603.000000,2681.788462,NaN,2024-10-05,2025-09-27,saturday,avg_from_daily
2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,1814.557692,1817.115385,NaN,2024-10-06,2025-09-28,sunday,avg_from_daily
3,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,5344.149798,5477.064777,NaN,2024-10-01,2025-09-30,weekday,avg_from_daily
4,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900209,19th Street / Oakland,37.808078,-122.268758,1905.000000,1979.714286,NaN,2024-10-14,2025-09-01,holiday,avg_from_daily


In [332]:
len(df_agg_combined["organization_name"].unique())

19

In [219]:
df_agg_combined.to_csv("fact_ridership_output/fct_ridership_2026-03-12_without_id.csv")

### Debug for missing data

In [333]:
len(df_agg_combined)

51329

In [334]:
t_df_agg = df_agg_combined.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False) \
    .size().sort_values(by="size", ascending=False)
t_df_agg[t_df_agg["size"] > 1]["organization_name"].unique()

array([], dtype=object)

In [335]:
t_df_stop_count = df_agg_combined.drop_duplicates(subset=["organization_name", "stop_id", "gtfs_stop_id", "stop_name"]).groupby("organization_name", as_index=False, dropna=False).size()
t_df_stop_count.sort_values(by="organization_name")

,organization_name,size
0,City and County of San Francisco,18
1,City of Culver City,379
2,City of Fresno,1618
3,City of Santa Monica,916
4,Foothill Transit,1711
5,Gold Coast Transit District,647
6,"Golden Gate Bridge, Highway and Transportation...",339
7,Long Beach Transit,1931
8,OmniTrans,1423
9,Orange County Transportation Authority,4818


**Note** agencies that have missing stops:
- City of Santa Monica 964 -> 916
- OmniTrans 1427 -> 1423

In [223]:
len(df_ridership[(df_ridership["organization_name"] == "City of Santa Monica")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

964

In [224]:
len(df_agg_combined[(df_agg_combined["organization_name"] == "City of Santa Monica")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

916

In [225]:
len(df_ridership[(df_ridership["organization_name"] == "City of Santa Monica") &
                 (df_ridership["start_date"] == "2025-08-01")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

916

In [342]:
len(df_ridership[(df_ridership["organization_name"] == "OmniTrans") & (df_ridership["start_date"] == "2024-07-01")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

1423

In [339]:
len(df_agg_combined[(df_agg_combined["organization_name"] == "OmniTrans")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

1423

In [340]:
len(df_ridership[(df_ridership["organization_name"] == "OmniTrans") &
                 (df_ridership["start_date"] == "2024-07-01")][["stop_name", "stop_id", "gtfs_stop_id"]].drop_duplicates())

1423

**Note:** In combined dataset, Cit of Santa Monica (Big Blue Bus) and OmniTrans have less number of stops in preprocessed/raw data, because we filtered only most recent time period.

#### Add key

In [343]:
def generate_key(*args, len=16, sep="|"):
    """Generate short reproducible hash ID based on input string"""
    key = sep.join(str(arg) for arg in args if arg is not None).encode("utf-8")
    return hashlib.sha1(key).hexdigest()[:len].upper()

In [344]:
df_agg_combined.head()

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,avg_daily_alightings,avg_daily_boardings,avg_daily_ridership,start_date,end_date,day_type,avg_daily_method
0,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2108.857143,2120.214286,NaN,2024-10-14,2025-09-01,holiday,avg_from_daily
1,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2603.000000,2681.788462,NaN,2024-10-05,2025-09-27,saturday,avg_from_daily
2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,1814.557692,1817.115385,NaN,2024-10-06,2025-09-28,sunday,avg_from_daily
3,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,5344.149798,5477.064777,NaN,2024-10-01,2025-09-30,weekday,avg_from_daily
4,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900209,19th Street / Oakland,37.808078,-122.268758,1905.000000,1979.714286,NaN,2024-10-14,2025-09-01,holiday,avg_from_daily


In [347]:
df_agg_combined["record_id"] = df_agg_combined.apply(lambda row: generate_key(row["organization_name"],
                                                           row["service_name"],
                                                           row["stop_id"],
                                                           row["gtfs_stop_id"],
                                                           row["stop_name"],
                                                           row["day_type"],
                                                           row["start_date"]
                                                          ), axis=1)

In [348]:
df_agg_combined.head()

,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,avg_daily_alightings,avg_daily_boardings,avg_daily_ridership,start_date,end_date,day_type,avg_daily_method,record_id
0,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2108.857143,2120.214286,NaN,2024-10-14,2025-09-01,holiday,avg_from_daily,E2833D7F62AA2DA2
1,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2603.000000,2681.788462,NaN,2024-10-05,2025-09-27,saturday,avg_from_daily,86650A6C9839CFD9
2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,1814.557692,1817.115385,NaN,2024-10-06,2025-09-28,sunday,avg_from_daily,B8443E0B431DE31E
3,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,5344.149798,5477.064777,NaN,2024-10-01,2025-09-30,weekday,avg_from_daily,1216F8AEF13AC9AB
4,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900209,19th Street / Oakland,37.808078,-122.268758,1905.000000,1979.714286,NaN,2024-10-14,2025-09-01,holiday,avg_from_daily,825C4F49C743979D


In [349]:
# check duplicates
t_df_agg = df_agg_combined.groupby("record_id").size().reset_index(name="count")
t_df_agg[t_df_agg["count"] > 1]

,record_id,count


In [350]:
df_agg_combined = df_agg_combined[["record_id", "dataset_id", "organization_name", "service_name", "stop_id", "gtfs_stop_id", "stop_name", "stop_lat", "stop_lon", 
                 "avg_daily_alightings", "avg_daily_boardings", "avg_daily_ridership", "avg_daily_method", "start_date", "end_date", "day_type"]]
df_agg_combined.head()

,record_id,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,avg_daily_alightings,avg_daily_boardings,avg_daily_ridership,avg_daily_method,start_date,end_date,day_type
0,E2833D7F62AA2DA2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2108.857143,2120.214286,NaN,avg_from_daily,2024-10-14,2025-09-01,holiday
1,86650A6C9839CFD9,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2603.000000,2681.788462,NaN,avg_from_daily,2024-10-05,2025-09-27,saturday
2,B8443E0B431DE31E,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,1814.557692,1817.115385,NaN,avg_from_daily,2024-10-06,2025-09-28,sunday
3,1216F8AEF13AC9AB,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,5344.149798,5477.064777,NaN,avg_from_daily,2024-10-01,2025-09-30,weekday
4,825C4F49C743979D,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900209,19th Street / Oakland,37.808078,-122.268758,1905.000000,1979.714286,NaN,avg_from_daily,2024-10-14,2025-09-01,holiday


#### Export

In [357]:
df_agg_combined[df_agg_combined["stop_id"].isna() & 
                df_agg_combined["gtfs_stop_id"].isna() &
                df_agg_combined["stop_name"].isna()]

,record_id,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,avg_daily_alightings,avg_daily_boardings,avg_daily_ridership,avg_daily_method,start_date,end_date,day_type
13112,510B091844DADD53,957BBF3AD8FC1B44,San Mateo County Transit District,SamTrans,<NA>,<NA>,NaN,NaN,NaN,0.0,0.0,NaN,avg_from_daily,NaN,NaN,weekend


In [351]:
today = date.today().strftime("%Y-%m-%d")
df_agg_combined.to_csv(f"fact_ridership_output/fct_ridership_{today}.csv", index=False) # save a copy

In [352]:
df_agg_combined.head()

,record_id,dataset_id,organization_name,service_name,stop_id,gtfs_stop_id,stop_name,stop_lat,stop_lon,avg_daily_alightings,avg_daily_boardings,avg_daily_ridership,avg_daily_method,start_date,end_date,day_type
0,E2833D7F62AA2DA2,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2108.857143,2120.214286,NaN,avg_from_daily,2024-10-14,2025-09-01,holiday
1,86650A6C9839CFD9,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,2603.000000,2681.788462,NaN,avg_from_daily,2024-10-05,2025-09-27,saturday
2,B8443E0B431DE31E,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,1814.557692,1817.115385,NaN,avg_from_daily,2024-10-06,2025-09-28,sunday
3,1216F8AEF13AC9AB,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900109,12th Street / Oakland City Center,37.803482,-122.271630,5344.149798,5477.064777,NaN,avg_from_daily,2024-10-01,2025-09-30,weekday
4,825C4F49C743979D,011CF30F49575609,San Francisco Bay Area Rapid Transit District,Bay Area Rapid Transit,<NA>,900209,19th Street / Oakland,37.808078,-122.268758,1905.000000,1979.714286,NaN,avg_from_daily,2024-10-14,2025-09-01,holiday


In [353]:
# check duplicates
t_dup_check = df_agg_combined.groupby(["organization_name", "stop_id", "gtfs_stop_id", "stop_name", "day_type"], as_index=False, dropna=False).size().sort_values(by="size", ascending=False)
t_dup_check[t_dup_check["size"] > 1]

,organization_name,stop_id,gtfs_stop_id,stop_name,day_type,size
